governance/api-reviews.md

```diff
@@ -200,3 +200,61 @@ When adding new ops, look for:
    it CPU, GPU, XLA, TPU, etc) and prefer to have at least a plan for writing
    device-agnostic code
  - should the python layer for this operation support raggedtensor/sparsetensor?
+
+## Experimental APIs
+
+Experimental APIs are APIs which have the word 'experimental' somewhere in their
+name; for example `tf.experimental.foo`, or `tf.foo.experimental.Bar`, or
+`tf.foo(experimental_bar=True)` or `tf.Foo().experimental_bar()`. We generally
+prefer experimental namespaces when possible, so prefer
+`tf.foo.experimental.Bar` over `tf.foo.ExperimentalBar`.
+
+Experimental APIs are APIs intended to be added to TensorFlow as-is, but which
+we reserve the right to change in backwards-incompatible ways if we have
+to.
+
+No temporary APIs should be added to experimental (i.e. "we just need this until
+certain bugfix or certain new feature becomes available" is not a valid reason
+to add an API with experimental in the name.) 
+
+No API with known deficiencies should be added to experimental. Experimental
+APIs should, to the best of our knowledge, not be expected to change in a known
+way (no argument with a known bad name, etc). Experimental can, however, be used
+for APIs which are a work-in-progress: it's fine to add experimental methods to
+a base class even if those methods are only implemented on some subclasses as
+long as we expect all classes to eventually implement those.
+
+The same amount of due diligence required for a real API is required for an
+experimental API: this means tests, benchmarks, documentation, end-to-end
+examples, etc
+
+Experimental APIs are not a license to break users. This means:
+ 1. we do not remove experimental APIs which are widely used without an effort
+    to help migrate users away
+ 2. experimental APIs are not removed without warning and don't have
```

Comment:
would this also cover graduation from experimental? Once API stabilised and was added to permanent namespace?

Reply:
non-experimental APIs cannot be removed (within a major version) already

Reply:
i meant graduation from experimental namespace to non-experimental

Reply:
graduation means removing the experimental API and adding a non-experimental equivalent; we cannot remove the experimental endpoint without plenty of warning

Reply:
yes, I guess it a good solution, additionally we might mention in warning message, that people can migrate to stable API

Reply:
Please remember that the Addons entry point could have [two graduation paths](See https://github.com/tensorflow/addons/blob/master/MIGRATION_TO_CORE.md#process) as Addons is (will still be?) the fusion of Keras and Tensorflow `contribs` communities and so we need to clarify what this means for the Tensorflow/Keras ecosystem.

We need to clarify how to route to one entry point or to the other one or pass through the community "Addons" before to be graduated and what means.

It is a little bit clear to me when you need to "patch" an experimental API in some experimental namespace but there are many edge cases. It is very foggy instead when you need to extend the experimental namespace with new user contributed features. Also I cannot figure out e.g. if a `tfa.image` populer ops could be graduated directly in `tf.image` or it will need to pass thought a `tf.experimental` stage.

Reply:
I think whether stuff can graduate directly from addons to tf or through experimental will be decided on a case-by-case basis

Reply:
Ok but it is still not clear to me if an external contributor need to contribute in `tf.exeperimental` or in `tfa.` Is this `tf.foo.experimetnal` policy reserved to Google employee?